libs/contrib/Data/Rational.idr

```diff
@@ -0,0 +1,91 @@
+module Data.Rational
+
+import Data.Nat.Factor
+
+%default total
+
+public export
+data Nat1 : Type where
+   S : Nat -> Nat1
+
+export
+Cast Nat1 Integer where
+   cast (S n) = (natToInteger n) + 1
+
+mult : Nat1 -> Nat1 -> Nat1
+mult (S m) (S n) = S (m + n + m * n)
+
+toNat_notbothzero : Nat1 -> Nat
+toNat_notbothzero (S n) = S n
+
+fromNat_coerce : Nat -> Nat1
+fromNat_coerce Z = S Z
+fromNat_coerce (S n) = S n
+
+public export
+record Rational where
+   constructor RatioOf
+   numerator : Integer
+   denominator : Nat1
+
+
+-- TODO: inline all this as prim__
```

Comment:
What do you mean by this? Are you going to resolve this todo before it is merged

Reply:
It needs backend support to be fast. Right now it "should" be working.